src/test/test-set.c

```diff
@@ -117,6 +117,36 @@ static void test_set_ensure_allocated(void) {
         assert_se(set_size(m) == 0);
 }
 
+static void test_set_copy(void) {
+        Set *s, *copy;
+        char *key1, *key2, *key3, *key4;
+
+        log_info("/* %s */", __func__);
+
+        key1 = strdup("key1");
+        assert_se(key1);
+        key2 = strdup("key2");
+        assert_se(key2);
+        key3 = strdup("key3");
+        assert_se(key3);
+        key4 = strdup("key4");
+        assert_se(key4);
+
+        s = set_new(&string_hash_ops);
+
+        set_put(s, key1);
+        set_put(s, key2);
+        set_put(s, key3);
+        set_put(s, key4);
+
+        copy = set_copy(s);
+
+        assert(set_equal(s, copy));
+
+        set_free_free(copy);
```

Comment:
you free the strings in the set with this even thouhg `s` also has a references to them still. Thus after this call completed `s` will have 4 entries pointing to invalid memory. that's ugly. please swap this line with the next, so that the strings are only freed when the final set object is freed.